Change `tagged` variable in Storage component to plural form.

Py0000 opened the pull request:
Fixes #147 

Change the variable `tagged` to `tags` in the affected class (`src/main/java/seedu/address/storage/JsonAdaptedPerson.java`).

yamidark commented:
According to this [issue](https://youtrack.jetbrains.com/issue/IDEA-189060) with some of the newer Intellij versions, even though the JAXB dependencies have been imported through gradle, and the code compiles and runs successfully, the IDE will still 'detect' and show code errors in these files (same problem occuring in AB4 as well). 
@pyokagan @yamgent @Zhiyuan-Amos Can help verify this on your side as well?

This will probably cause confusion to the students. Should we add a note of this error in the setting up of project in DeveloperGuide, and a way to remove the code error warnings?

![untitled](https://user-images.githubusercontent.com/18352498/43989458-f2c82b9c-9d7c-11e8-9093-c8baec4c9bb3.png)

yamidark commented:
Hmm, seems to just be an issue with `Ultimate Edition`, tried `Standard Edition` and there were no errors.
Will add a "Troubleshooting" section at the bottom just in case.

Zhiyuan-Amos commented on docs/DeveloperGuide.adoc:
```diff
@@ -26,6 +26,11 @@
+=== Troubleshooting
+
+* Problem: Intellij is reporting code errors in `StorageFile`, but the project can be build and run successfully. +
+Solution: Place your cursor over any of the code section highlighted in red. Press ALT + ENTER, and select `Add '--add-modules java.xml.bind' to module compile options`. The code errors should no longer appear.
```
Should ALT + ENTER be codified i.e. `ALT + ENTER`?

yamidark replied:
It wasn't codified in AB2 DG, so guess don't need to here

Zhiyuan-Amos commented on docs/DeveloperGuide.adoc:
```diff
@@ -26,6 +26,11 @@
+=== Troubleshooting
+
+* Problem: Intellij is reporting code errors in `StorageFile`, but the project can be build and run successfully. +
+Solution: Place your cursor over any of the code section highlighted in red. Press ALT + ENTER, and select `Add '--add-modules java.xml.bind' to module compile options`. The code errors should no longer appear.
```
Should we format this according to [AB-Level 4](https://github.com/se-edu/addressbook-level4/blob/master/docs/DeveloperGuide.adoc#53-troubleshooting-testing)?

yamidark commented on docs/DeveloperGuide.adoc:
```diff
@@ -9,23 +9,29 @@
+
+* Reason: This is an ongoing https://youtrack.jetbrains.com/oauth?state=%2Fissue%2FIDEA-189060[issue] with some of the newer versions of IntelliJ.
+* Solution: Place your cursor over any of the code section highlighted in red. Press ALT + ENTER, and select `Add '--add-modules java.xml.bind' to module compile options`. The code errors should no longer appear.
+
```
Hmm, we can ask them to check for this in the last step, as this shouldn't be a solution to failing builds/tests, if the project was imported correctly as a gradle project.

damithc replied:
Sure.

yamgent commented:
> This will probably cause confusion to the students. Should we add a note of this error in the setting up of project in DeveloperGuide, and a way to remove the code error warnings?

While porting the tutorial in https://github.com/se-edu/se-book/pull/62, I spotted the same error. Instead of the <kbd>Alt</kbd>+<kbd>Enter</kbd> trick, I used `module-info.java` to solve the issue.

Strangely enough, <kbd>Alt</kbd>+<kbd>Enter</kbd> does not seem to suggest modifying `module-info.java`. Maybe at this stage not many people use modules, thus that recommendation might not have been implemented yet 🤔.

yamidark commented:
>Strangely enough, Alt+Enter does not seem to suggest modifying module-info.java. Maybe at this stage not many people use modules, thus that recommendation might not have been implemented yet 

Tried adding an empty `module-info.java`, and the <kbd>Alt</kbd> + <kbd>Enter</kbd> does suggest adding this to `module-info.java` (on my side).

However, adding the `module-info.java` with all the required dependencies instead cause the program to crash with:
`Package seedu.addressbook.storage.jaxb with JAXB class seedu.addressbook.storage.jaxb.AdaptedAddressBook defined in a module addressbook.level3 must be open to at least java.xml.bind module.`

@yamgent Does your code build and run with the added `module-info.java`?

yamgent commented:
> Package seedu.addressbook.storage.jaxb with JAXB class seedu.addressbook.storage.jaxb.AdaptedAddressBook defined in a module addressbook.level3 must be open to at least java.xml.bind module.

This means that it expects an `opens` statement for `seedu.addressbook.storage.jaxb`:

```java
module addressbook.level3.main {
    requires java.xml.bind;
    requires javafx.fxml;
    requires javafx.graphics;
    requires javafx.controls;

    opens seedu.addressbook.storage.jaxb to java.xml.bind;
}
```

I tried fixing the errors but eventually end up in a runtime error (can be traced to some JAXB problem, seems like that is not getting exported properly? maybe that has to do with the errors that we are getting on Travis?).

If we are running out of time, the <kbd>Alt</kbd>+<kbd>Enter</kbd> can be an acceptable alternative.

yamidark commented:
>I tried fixing the errors but eventually end up in a runtime error (can be traced to some JAXB problem, seems like that is not getting exported properly

I tried debugging this too, seems like its the same error as I mentioned in https://github.com/se-edu/addressbook-level3/commit/b72eec207cef33fe279262f69fe290b84ac581eb, for some reason it is unable to get the resource `main.window.fxml`. Currently trying to find a way to resolve this...

yamidark commented:
@yamgent 
Managed to somewhat find the problem with this...

So when a `-module-info.java` is specified, IntelliJ will run using the 'modulepath' instead of 'classpath', which is set to `out/production/classes`. However, gradle places the resources in `out/production/resources`, thus the program is unable to find the resources and returns `null` for `Class.getResource`. 

There was also a need to `opens` the 'ui' package to `javafx.fxml` so that the `MainWindow` becomes accessible, and after that changing the location of the `Darktheme.css` in `mainwindow.fxml` to `@Darktheme.css` instead (since it searched relative to the opened 'ui' package instead?).

However, I couldn't really get to specify the output of the resources to be in the same module path, even after specifying `output.resourceDirs` in gradle, so I'm still somewhat stuck here.

pyokagan commented:
Gradle doesn't have first-class support for Java modules (development seems to be slow upstream as well). I suppose trying to convert the project to use Java modules without its support would be an uphill battle.

pyokagan commented:
We could try using https://github.com/zyxist/chainsaw for fun, just to see what happens, but I don't think the results should be committed into master.

pyokagan commented:
> However, gradle places the resources in out/production/resources, thus the program is unable to find the resources and returns null for Class.getResource.

This statement seems weird, btw. Gradle is writing to `out/`??? I thought only IntellIJ did that.

yamidark commented:
>This statement seems weird, btw. Gradle is writing to out/??? I thought only IntellIJ did that.

Whoops, sorry, meant to say IntelliJ does that when we import as a gradle project 💦.

yamgent commented:
@yamidark If you go to `File` -> `Project Structure` -> `Modules` -> `addressbook-level3_main` -> `Paths`, and select "Inherit project compile output path", does it work?

yamidark commented:
>If you go to File -> Project Structure -> Modules -> addressbook-level3_main -> Paths, and select "Inherit project compile output path", does it work?

Yep, it works! 👍.

pyokagan commented:
@yamgent One thing before you proceed with this though: make sure that the same approach works for AB-4 (and whether e.g. `./gradlew run` still works or not over there once there is a `module-info.java` file).

yamidark commented:
@yamgent @pyokagan 

Managed to configure the `module-info.java` and `build.gradle` get it to run and all the tests to pass in AB-4 along with gradle, but IntelliJ still has some issues.

In IntelliJ, there was still a runtime error when trying to run the app, 
`impl.org.controlsfx.version.VersionChecker (in module controlsfx) cannot access class com.sun.javafx.runtime.VersionInfo (in module javafx.base) because module javafx.base does not export com.sun.javafx.runtime to module controlsfx`.
The only solution I managed to find for this would be to add 
`--add-exports=javafx.base/com.sun.javafx.runtime=controlsfx` to the VM arguments.

Also, the test classes will still have the same problem where `java.xml.bind` has code errors (although we can run tests normally), since it doesn't have a `module-info.java`, and I'm unable to add it in because of [split packages](https://www.logicbig.com/tutorials/core-java-tutorial/modules/split-packages.html).

From this, I believe that trying to use Java modules compared to just <kbd>Alt</kbd> + <kbd>Enter</kbd> to resolve this issue is really overkill and also too troublesome for the students.

yamgent commented:
> From this, I believe that trying to use Java modules compared to just Alt + Enter to resolve this issue is really overkill and also too troublesome for the students.

Ok, let's just use <kbd>Alt</kbd>+<kbd>Enter</kbd> then, it is still in some sense using the module system, while still providing a simple fix for the problem.

pyokagan commented:
I agree it would be ideal if the ABs could be migrated to the module system, but it seems that the ecosystem has not caught up yet.

@yamidark Maybe you could submit a PR with your work to AB-4? Might be useful in the future when it's time to migrate.  If your commits are polished enough we could possibly just use your work as-is and keep your authorship, so your efforts won't get wasted.

pyokagan commented:
@yamidark 

> ```
> Java 9 introduces a module system, causing internal packages such as
> javax.* and com.sun.* packages to be inaccessible[1].
> 
> ...
> 
> [1]: http://openjdk.java.net/jeps/260
> ```

From my understanding, it is more to do with the fact that (1) AB-3 is an unnamed module since it does not have a `module-info.java`, and (2) the `java.xml.bind` module is not part of the default set of root modules for unnamed modules?

So, http://openjdk.java.net/jeps/261 would be a better link? Specifically since it mentions:

> The default set of root modules for the unnamed module is based upon the java.se module rather than the java.se.ee module. Thus, by default, code in the unnamed module will not have access to APIs in the following modules:
> 
> ```
> java.activation
> java.corba
> java.transaction
> java.xml.bind
> java.xml.ws
> java.xml.ws.annotation
> ```

yamidark commented:
@pyokagan 
>From my understanding, it is more to do with the fact that (1) AB-3 is an unnamed module since it does not have a module-info.java, and (2) the java.xml.bind module is not part of the default set of root modules for unnamed modules?
So, http://openjdk.java.net/jeps/261 would be a better link?

Sounds good 👍. Will edit the message.

pyokagan commented:
@yamidark 

> ```
> Java 9 introduces a module system, causing internal packages such as
> java.xml.bind to be inaccessible to unnamed modules by default[1]. As
> ```


Just to check again: is it really true that `java.xml.bind` is an internal API? Or is it just not accessible by default just because it is a Java EE package?

http://openjdk.java.net/jeps/261 says that "Modules that define Java EE APIs, or APIs primarily of interest to Java EE applications, have been deprecated and will be removed in a future release. They are not resolved by default for code on the class path", so I'm leaning towards the latter. I don't really think it has anything to do with internal APIs as well.

(Because, if we are really using an internal API, I think we have a bigger problem ;-) )


Also, a nit: http://openjdk.java.net/jeps/261 says that:

"""This is an intentional, if painful, choice, driven by two goals:

* To avoid unnecessary conflicts with popular libraries that define types in some of the same packages. The widely-used jsr305.jar, e.g., defines annotation types in the javax.annotation package, which is also defined by the java.xml.ws.annotation module.

* To make it easier for existing application servers to migrate to JDK 9. Application servers often override the content of one or more of these modules, and in the near term they are most likely to do so by continuing to place the necessary non-modular JAR files on the class path. If these modules were resolved by default then the maintainers of application servers would have to take awkward actions to exclude them in order to override them."""

So I'd say this is a decision made independently of the module system (i.e. the module system doesn't directly "cause" this, it is possible for the `java.se.ee` module to be part of the root set of modules, it is just that they explicitly chose not to), so we shouldn't say it is `causing`?

yamidark commented:
@pyokagan 

>Just to check again: is it really true that java.xml.bind is an internal API? Or is it just not accessible by default just because it is a Java EE package?
http://openjdk.java.net/jeps/261 says that "Modules that define Java EE APIs, or APIs primarily of interest to Java EE applications, have been deprecated and will be removed in a future release. They are not resolved by default for code on the class path", so I'm leaning towards the latter. I don't really think it has anything to do with internal APIs as well.

Yes, your right. Seems I mixed up public and internal APIs together. Will update my commit message.

>So I'd say this is a decision made independently of the module system (i.e. the module system doesn't directly "cause" this, it is possible for the java.se.ee module to be part of the root set of modules, it is just that they explicitly chose not to), so we shouldn't say it is causing?

True, seems I picked the wrong choice of words 😓.

pyokagan commented:
@yamidark Thanks for the clarification.

The new Java module system is quite complex, so the clarity is needed to ensure we are fixing the right thing :-)

pyokagan commented on docs/DeveloperGuide.adoc:
```diff
@@ -25,6 +26,9 @@
 . Run the `seedu.addressbook.Main` class (right-click the `Main` class and click `Run Main.main()`) and try executing a few commands
 . Run all the tests (right-click the `test` folder, and click `Run 'All Tests'`) and ensure that they pass
+. Open the `StorageFile` file and check for any code errors
+.. Due to an ongoing https://youtrack.jetbrains.com/oauth?state=%2Fissue%2FIDEA-189060[issue] with some of the newer versions of IntelliJ, code errors may be detected even if the project can be built and run successfully
```
The link should be https://youtrack.jetbrains.com/issue/IDEA-189060 . "oauth" seems to be some kind of internal authorization URL, and so would be less likely to have any guarantees of permanence.

Same for the commit message.

Py0000 committed:
Change `tagged` variable in Storage component to plural form.

Py0000 committed:
Fix Failing testcase after changes made to `tagged` variable.

Testcase `toModelType_typicalPersonsFile_success` fails

Revert the tag JsonProperty back to `tagged`

To follow the format used in `data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json`

damithc commented:
@Py0000 thanks for this PR, and sorry for the delay in processing it. Can you follow the instructions in https://github.com/se-edu/addressbook-level3/pull/148#issuecomment-1288977823 so that the PR is compliant with our process.

zhongfu committed:
Fix typo in DevOps guide testes -> tests (#144)

dependabot[bot] committed:
Bump nokogiri from 1.13.6 to 1.13.9 in /docs (#146)

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.13.6 to 1.13.9.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.13.6...v1.13.9)

---
updated-dependencies:
- dependency-name: nokogiri
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Py0000 committed:
Fix typo in custom-styles.scss order -> border (#150)

order-radius -> border-radius

dependabot[bot] committed:
Bump nokogiri from 1.13.9 to 1.13.10 in /docs (#151)

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.13.9 to 1.13.10.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.13.9...v1.13.10)

---
updated-dependencies:
- dependency-name: nokogiri
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

damithc committed:
UG: Ask to run the jar file using the command line

Double-clicking the jar file to run it can cause unexpected
behaviors in some cases (e.g., the app using a different JDK
version than the intended one).

Let's update the UG to ask users to use the 'java jar' command
by default.

damithc committed:
UG: Make example commands formatting PDF-friendly

Some example commands in the UG has the first word in bold font,
which requires the command to be formatted as two separate strings,
which in turn introduces extra whitespace in the middle of the
command when converting to PDF.

Let's remove the bold formatting so that the whole command can
be formatted as a single string.

damithc committed:
DG: Remove incorrect composition symbol

BetterModelClassDiagram shows the UniqueTagList as 'composed of'
Tag objects. However, this is an aggregation rather than a
composition.

As we don't usually show aggregation symbols in class diagrams,
let's remove the offending composition symbol altogether.

damithc committed:
DevOps.md: Remove extra '<'

damithc committed:
DevOps.md: Mention how to relax Codecov requirement

As Codecov service can sometimes run into operational problems
due to issues on their side, let's mention how to configure
CI to pass even if Codecov fails.

Py0000 committed:
Merge branch 'master' of https://github.com/Py0000/addressbook-level3-techteamversion

* 'master' of https://github.com/Py0000/addressbook-level3-techteamversion:
  DevOps.md: Mention how to relax Codecov requirement
  DevOps.md: Remove extra '<'
  DG: Remove incorrect composition symbol
  UG: Make example commands formatting PDF-friendly
  UG: Ask to run the jar file using the command line
  Bump nokogiri from 1.13.9 to 1.13.10 in /docs (#151)
  Fix typo in custom-styles.scss order -> border (#150)
  Bump nokogiri from 1.13.6 to 1.13.9 in /docs (#146)
  Fix typo in DevOps guide testes -> tests (#144)
